File: Medium/36. Valid Sudoku/solution.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        for(int i = 0; i < 10; i++){
            if(!IsValidRow(i, board) || !IsValidCol(i, board)) return false;
        }

        for(int i = 0; i < 9; i += 3){
            for(int j = 0; j < 9; j += 3){
                if(!IsValidUnit(i, j, board)) return false;
            }
        }

        return true;
    }

    public bool IsValidUnit(int x, int y, char[][] board) {
        int[] appear = new int[10];
        for(int i = x; i < x + 3; i++){
            for(int j = y; j < y + 3; j++){
                if(board[i][j] != '.'){
                    int num = board[i][j] - '0';
                    appear[num] += 1;
                    if(appear[num] != 1) return false;
                }
            }
        }

        return true;
    }

    public bool IsValidRow(int x, char[][] board) {
        int[] appear = new int[10];
        for(int i = 0; i < 9; i++){
            if(board[x][i] != '.'){
                int num = board[x][i] - '0';
                appear[num] += 1;
                if(appear[num] != 1) return false;
            }
        }

        return true;
    }

    public bool IsValidCol(int y, char[][] board) {
        int[] appear = new int[10];
        for(int i = 0; i < 9; i++){
            if(board[i][y] != '.'){
                int num = board[i][y] - '0';
                appear[num] += 1;
                if(appear[num] != 1) return false;
            }
        }

        return true;
    }
}

File: Medium/36. Valid Sudoku/solution.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for i in range(9):
            if self.isValidRow(i, board) == False:
                return False
            if self.isValidCol(i, board) == False:
                return False

        for i in range(0, 7, 3):
            for j in range(0, 7, 3):
                if self.isValidUnit(i, j, board) == False:
                    return False

        return True

    def isValidUnit(self, x, y, board):
        appear = [0] * 10
        for i in range(x, x + 3):
            for j in range(y, y + 3):
                if board[i][j] != ".":
                    appear[int(board[i][j])] += 1
                    if appear[int(board[i][j])] != 1:
                        return False
        return True

    def isValidRow(self, x, board):
        appear = [0] * 10
        for i in range(9):
            if board[x][i] != ".":
                appear[int(board[x][i])] += 1
                if appear[int(board[x][i])] != 1:
                    return False
        return True

    def isValidCol(self, y, board):
        appear = [0] * 10
        for i in range(9):
            if board[i][y] != ".":
                appear[int(board[i][y])] += 1
                if appear[int(board[i][y])] != 1:
                    return False
        return True
